### PR TITLE
[https://nvbugs/5764627][fix] Fix generation logits with streaming and improve runtime of logits testcase. Also fixes https://nvbugs/5573238

### PR DESCRIPTION
**Pull request by stnie**
## Release Notes

* **Bug Fixes**
  * Fixed race conditions in generation logits tracking during streaming operations.
  * Optimized memory efficiency in streaming mode by eliminating unnecessary deep copy operations.

* **Tests**
  * Re-enabled previously skipped tests for logits and log probabilities functionality.
  * Enhanced test coverage and validation for streaming scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

## Description

test_return_logits.py is too slow. 
- Too many llm Instances were created. This PR resolves this by changing some fixtures to parametrizations. The adjusted test only creates a total of 4 llm instances.
- test_generate_with_return_logits and test_generate_async_with_return_logits have been consolidated into a single testcase to better reuse llm fixtures and reduce code duplication.
- Previous xfail cases could cause the llm shutdown to hang, which resulted in the testcase waiting for timeout (120s). This PR removes the xfail cases.
- Bug 5573238 is resolved, by creating a deep copy of the LogitsStorage when creating the response, to prevent a race condition 

## Test Coverage

test_return_logits.py

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by Funatiq — APPROVED** [review]

**Review by Funatiq — APPROVED** [review]

**Review by yibinl-nvidia — APPROVED** [review]

**Review by ixlmar — COMMENTED** [review]
Commented on test changes

File: tests/unittest/_torch/sampler/test_return_logits.py [hunk 1]
```diff
@@ -193,47 +159,41 @@ def test_generate_async_with_return_logits(
...
-            if return_log_probs:
-                assert len(sequence.logprobs) == idx + 1
-            else:
-                assert len(sequence.logprobs) == 0
+                reuse_cache=reuse_cache,
+                return_log_probs=return_log_probs,
+                idx=idx,
+                output=output,
+                streaming=True)
+    else:
```
[the remark this answers is not in the document]
Comment (stnie) [reply tests/unittest/_torch/sampler/test_return_logits.py]: I will add this as an additional check

File: tests/unittest/_torch/sampler/test_return_logits.py [hunk 2]
```diff
@@ -193,47 +159,41 @@ def test_generate_async_with_return_logits(
...
+                )):
+            check_generated_output(
+                gather_context_logits=gather_context_logits,
+                gather_generation_logits=gather_generation_logits,
+                sampling_params=sampling_params,
+                reuse_cache=reuse_cache,
+                return_log_probs=return_log_probs,
+                idx=idx,
+                output=output,
+                streaming=False)
```
[the remark this answers is not in the document]
Comment (stnie) [reply tests/unittest/_torch/sampler/test_return_logits.py]: See previous answer

File: tests/unittest/_torch/sampler/test_return_logits.py [hunk 3]
```diff
@@ -102,89 +90,67 @@ def _exit_with_xfail_on_timeout(self, exc_type, exc_value,
...
+                           output, streaming):
+    if gather_context_logits:
+        assert output.context_logits is not None
+        # NOTE: prompt_token_ids of "A B C" becomes [1, 319, 350, 315]
+        expected_len = len(prompts[0].split()) + 1
+        if reuse_cache:
+            # Either all tokens are calculated (expected_len) or all tokens are reused (1)
+            assert (output.context_logits.shape[0]
+                    == expected_len) or (output.context_logits.shape[0] == 1)
+            # "Known bug: https://nvbugs/5577178" Reuse may cause the context logits to be shorter than expected.
```
[the remark this answers is not in the document]
Comment (stnie) [reply tests/unittest/_torch/sampler/test_return_logits.py]: From what I understood this is expected behavior, which is why I removed the `xfail.`

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 4]
```diff
@@ -2119,6 +2126,16 @@ def _executor_loop_overlap(self):
                 else:
                     self._enqueue_responses([])
 
+                # Call set_exclude_last_generation_logits after _process_previous_batch.
+                # If set before, the response of a request may be incorrect, as it will
+                # use the wrong indices for generation logits when streaming is enabled.
+                if can_queue:
```
[the remark this answers is not in the document]
Comment (stnie) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: I created a function which performs this check. As this update is slightly different for each type of executor_loop I decided to directly call it in the corresponding loops

**Review by Funatiq — APPROVED** [review]
